source/darwin/dispatch_queue_event_loop.c

```diff
@@ -0,0 +1,698 @@
+
+    if (!begin_iteration(entry)) {
+        goto iteration_done;
+    }
+
+    struct dispatch_loop *dispatch_loop = entry->dispatch_queue_context->io_dispatch_loop;
+    // swap the cross-thread tasks into task-local data
+    struct aws_linked_list local_cross_thread_tasks;
+    aws_linked_list_init(&local_cross_thread_tasks);
+    s_lock_cross_thread_data(dispatch_loop);
+    dispatch_loop->synced_data.current_thread_id = aws_thread_current_thread_id();
+    dispatch_loop->synced_data.is_executing = true;
+    aws_linked_list_swap_contents(&dispatch_loop->synced_data.cross_thread_tasks, &local_cross_thread_tasks);
+    s_unlock_cross_thread_data(dispatch_loop);
+
+    aws_event_loop_register_tick_start(dispatch_loop->base_loop);
+
+    // run the full iteration here: local cross-thread tasks
+    while (!aws_linked_list_empty(&local_cross_thread_tasks)) {
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&local_cross_thread_tasks);
+        struct aws_task *task = AWS_CONTAINER_OF(node, struct aws_task, node);
+
+        /* Timestamp 0 is used to denote "now" tasks */
+        if (task->timestamp == 0) {
+            aws_task_scheduler_schedule_now(&dispatch_loop->scheduler, task);
+        } else {
+            aws_task_scheduler_schedule_future(&dispatch_loop->scheduler, task, task->timestamp);
+        }
+    }
+
+    // run all scheduled tasks
+    uint64_t now_ns = 0;
+    aws_event_loop_current_clock_time(dispatch_loop->base_loop, &now_ns);
+    aws_task_scheduler_run_all(&dispatch_loop->scheduler, now_ns);
+    aws_event_loop_register_tick_end(dispatch_loop->base_loop);
+
+    end_iteration(entry);
+
+iteration_done:
+    s_lock_service_entries(dispatch_queue_context);
+    s_scheduled_service_entry_destroy(dispatch_queue_context->scheduling_state, entry);
+    s_unlock_service_entries(dispatch_queue_context);
+    s_runlock_dispatch_loop_context(dispatch_queue_context);
+    s_release_dispatch_loop_context(dispatch_queue_context);
+}
+
+/**
+ * Checks if a new iteration task needs to be scheduled, given a target timestamp. If so, submits an iteration task to
+ * dispatch queue and registers the pending execution in the event loop's list of scheduled_services.
+ *
+ * If timestamp==0, the function will always schedule a new iteration as long as the event loop is not suspended.
+ *
+ * The function should be wrapped with the following locks:
+ *      dispatch_loop->context->lock: To retain the dispatch loop
+ *      dispatch_loop->synced_data.lock : To verify if the dispatch loop is suspended
+ *      dispatch_loop_context->scheduling_state->services_lock: To modify the scheduled_services list
+ */
+static void s_try_schedule_new_iteration(struct dispatch_loop_context *dispatch_loop_context, uint64_t timestamp) {
+    struct dispatch_loop *dispatch_loop = dispatch_loop_context->io_dispatch_loop;
+    if (!dispatch_loop || dispatch_loop->synced_data.suspended) {
```

Comment:
Not an issue if we remove suspended but if we are accessing synced_data.suspended, we should probably do so while holding the synced_data lock.